src/box/ck_constraint.c

```diff
@@ -200,7 +200,8 @@ ck_constraint_on_replace_trigger(struct trigger *trigger, void *event)
 			 "field_ref");
 		return -1;
 	}
-	vdbe_field_ref_prepare_tuple(field_ref, new_tuple);
+	vdbe_field_ref_prepare_tuple(field_ref, space->def->field_count,
+				     new_tuple);
```

Comment:
I don't understand. Why does it have to so complicated? How does `select *` fetch all fields? Why does it even go beyond `space->def->field_count` that you have to limit it here? Does it try to return all fields looking only at `ref->field_count`?

Reply:
Hi! Thank you for the review! The entire `struct vdbe_field_ref` is used for optimization and was introduced in commit https://github.com/tarantool/tarantool/commit/d24b3165904a69738c64456ce7ed05f06a8ae3f1. The `allocateCursor()` function allocates enough memory for `slots` that remember the shift of some fields of tuple. It is worth noting that the memory for cursor (`struct BtCursor`) is also allocated simultaneously and immediately after `slots`. In the `vdbe_field_ref_create()` function, which is called from `vdbe_field_ref_prepare_tuple()` and `vdbe_field_ref_prepare_data()`, the memory of the  `slots` is `memset()` to 0. However, the number of fields was determined from the number of fields of `msgpack`, and may be not the actual number of named fields. Thus, prior to https://github.com/tarantool/tarantool/commit/89057a21b5c1bb864f9f61adabd76b25af846a2c in case the number of named fields was less than the number of fields in `msgpack`, the `memset()` cursor was also `memset()` 0. However, in that patch, I used the number `tuple_format(tuple)->total_field_count`, which, as I found out yesterday, can also be greater than `space->def->field_count`.

The `field_count` field of the `struct vdbe_field_ref` structure must be equal to the minimum number of named fields and the number of fields in `msgpack`. It cannot be greater than the number of named fields, since in this case we will set the `memset()` cursor to 0, and it cannot be greater than the number of fields in `msgpack`, since in this case we can try to get a field that is not in `msgpack`.

According to the current logic, we cannot drop `memset()`. However, we can think of using the `slot_bitmask` field, so we don't need `memset()`. Do you think it's worth doing it?

Reply:
Ok, I see. Thanks for the detailed explanation. It feels like we miss an important step: actual creation of a valid `vdbe_field_ref` when it is allocated. `allocateCursor()` allocates it and fills with zeros using `memset()`. But the fact is that already in `allocateCursor()` you know the max number of fields and slot count - it is `nField` parameter. The problem here is that `vdbe_field_ref` at creation doesn't save its capacity and hence you have to somehow pass same or smaller field count to `prepare_tuple/data()`.

With this implementation we can't even add assertions that we do not access slots out of borders.

I suggest to do a bit more generic rework:
- There is a new field `vdbe_field_ref.field_capacity`. `allocateCursor()` should correctly create `vdbe_field_ref` not just with `memset(0)`, but do `ref.field_capacity = nField`;
- `vdbe_field_ref_prepare_data()/_tuple()` remains like now except that it does `ref.field_count = MIN(mp_field_count, ref.field_capacity)`.

That way you don't need to pass space/format field count to `prepare()` functions and it gets safer, because the capacity is always known.

As for bitmask to avoid `memset()` - I don't mind, didn't think of that. Could be done separately as an optimization.